experimental/compactdecimal/src/format.rs

```diff
@@ -0,0 +1,74 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use alloc::borrow::Cow;
+use fixed_decimal::{CompactDecimal, FixedDecimal};
+use writeable::Writeable;
+use zerovec::maps::ZeroMap2dCursor;
+
+use crate::compactdecimal::CompactDecimalFormatter;
+use crate::provider::{Count, PatternULE};
+
+/// An intermediate structure returned by [`CompactDecimalFormatter`](crate::CompactDecimalFormatter).
+/// Use [`Writeable`][Writeable] to render the formatted decimal to a string or buffer.
+pub struct FormattedCompactDecimal<'l> {
+    pub(crate) formatter: &'l CompactDecimalFormatter,
+    pub(crate) value: Cow<'l, CompactDecimal>,
+    pub(crate) plural_map: Option<ZeroMap2dCursor<'l, 'l, i8, Count, PatternULE>>,
+}
+
+impl<'l> Writeable for FormattedCompactDecimal<'l> {
+    fn write_to<W>(&self, sink: &mut W) -> core::result::Result<(), core::fmt::Error>
+    where
+        W: core::fmt::Write + ?Sized,
+    {
+        if self.value.exponent() == 0 {
+            self.formatter
+                .fixed_decimal_format
+                .format(self.value.significand())
+                .write_to(sink)
+        } else {
+            let plural_map = self.plural_map.as_ref().ok_or(core::fmt::Error)?;
+            let chosen_pattern = (|| {
+                if self.value.significand() == &FixedDecimal::from(1) {
+                    if let Some(pattern) = plural_map.get1(&Count::Explicit1) {
+                        return Some(pattern);
+                    }
+                }
+                let plural_category = self
+                    .formatter
+                    .plural_rules
+                    .category_for(self.value.significand());
+                plural_map
+                    .get1(&plural_category.into())
+                    .or_else(|| plural_map.get1(&Count::Other))
+            })()
+            .ok_or(core::fmt::Error)?;
+            match chosen_pattern.index {
+                u8::MAX => sink.write_str(&chosen_pattern.literal_text),
+                _ => {
+                    let i = usize::from(chosen_pattern.index);
+                    sink.write_str(
+                        chosen_pattern
+                            .literal_text
+                            .get(..i)
+                            .ok_or(core::fmt::Error)?,
+                    )?;
+                    self.formatter
+                        .fixed_decimal_format
+                        .format(self.value.significand())
+                        .write_to(sink)?;
```

Comment:
Issue: This might not work as expected with negative numbers. Please add some test cases for that (can be done after landing this PR).

The way FixedDecimalFormatter deals with signs (plus and minus) is to have 3 versions of the pattern that are selected at runtime: no sign, plus sign, and minus sign. So, we don't actually render a sign at runtime; we just select the pre-rendered pattern. I don't know if that's how we want to handle it in CDF since we have a lot more patterns and would like to avoid bloating the data.

Reply:
Added some examples in the docs tests; it works as I expect at least.

French `mille` is messy (though it doesn’t do something truly misleading like dropping the sign, it falls to that weird `one` case, thus `-1 millier`, which may be the least broken thing it can do; getting `-mille` would probably be tricky, and that looks utterly weird anyway).

Reply:
Yes, but try this in a locale where the sign and the compact affix are on the same side. For example:

https://github.com/unicode-org/cldr-json/blob/main/cldr-json/cldr-numbers-full/main/sw-KE/numbers.json

We have a pattern

```
              "1000000-count-other": "M0",
```

The correct output according to CLDR would be "+M3", but your code produces "M+3"

Reply:
In that same file, there is

```
              "1000-count-one": "elfu 0;elfu -0",
```

which demonstrates how CLDR does actually want us to consider these as proper decimal patterns that could have different signed and unsigned versions.

Reply:
(argh, too many threads.)

Yes, but there is also `elfu 0` without that dance, which is therefore suspect; and the millions don’t do that but do it in other variants of Swahili, etc.

I suspect we have bad data, but that having the sign next to the number is the better default, since many swahili patterns go out of their way to stick it back there.

Reply:
> "+M3" is likely a reasonable and correct outcome

I am unconvinced of that, given that it abbreviates something that seems to want to be `milioni +3`.

Bear in mind that while compact decimal is used quite a bit more than scientific, its edge cases (including the use of negative numbers) quickly get uncharted (after all, the Romance languages are only just starting not to be a flagrantly ungrammatical mess, and that is for positive values).

We will indeed need to deal with that for currencies (along with a number of other complexities).

Reply:
We have some evidence from the number range work that designers have special requirements when it comes to single-character symbols. They tend to prefer single characters to be "glued" to the number, with extra annotations (like range separators and signs) displayed outside of the single-character symbol.

Reply:
It's unambiguous that the current UTS 35 requires "+M3" instead of "M+3". Perhaps that data is wrong, but even if it is, there will continually be more cases where we need to handle cases like this. Should we appeal to CLDR to get clarity?

Reply:
Either way I need to open a CLDR ticket because it is clear that there is a lot of wrongness in the Swahili data, for instance 
https://github.com/unicode-org/cldr-json/blob/f27f55f1dd487af7cf4260f56296ee1c7649b7fc/cldr-json/cldr-numbers-full/main/sw-KE/numbers.json#L34-L36
next to https://github.com/unicode-org/cldr-json/blob/f27f55f1dd487af7cf4260f56296ee1c7649b7fc/cldr-json/cldr-numbers-full/main/sw-KE/numbers.json#L62-L64.

If it turns out that something like `sign` `prefix` `number` etc. is needed for compact decimal (as opposed to compact currency) we can add a bit that says « sign goes at the beginning », but I suspect many of the patterns that do that (again, for _compact decimal_, not compact currency or ranges etc.) do that because CLDR hands the translators a handy footgun.

Reply:
> If it turns out that something like `sign` `prefix` `number` etc. is needed for compact decimal (as opposed to compact currency) we can add a bit that says « sign goes at the beginning »

OK. Having only two valid places to put the sign (inside or outside the affix) is probably reasonable, but it's yet another invariant that we'll need to enforce in datagen. It does make things a bit easier and smaller.